### PR TITLE
fix js error on initial toast load

### PR DESCRIPTION
**Pull request by iszmais**
https://mantis.ilias.de/view.php?id=34590

This PR mostly reverts changes made in https://github.com/ILIAS-eLearning/ILIAS/pull/5012 and add all toasts that are preexisting to the toast container withing the pagebuilder to prevent an additional request.

### CODE REVIEW

**Comment by iszmais** [comment]
@lscharmer @Amstutz @klees Pls leave your opinion on this.
I wouldnt see the need to take that to a JF until the is no major discussion point.
Feel free to open such.

Greeting,
@iszmais

**Comment by klees** [comment]
Hi @alex40724,

you are listed as implicit maintainer for Services/Notification, hence I assigned you. Please reassign if there is a person you deem more suitable.

Thanks!

**Review by klees — APPROVED** [review]
I'm not much into the whole feature and problem, but this looks plausible indeed.

**Comment by mjansenDatabay** [comment]
It is `Services/Notifications`, so I deassigned @alex40724 and assigned @mjansenDatabay

**Comment by klees** [comment]
Oh, that's an intricate distinction =)

**Comment by mjansenDatabay** [comment]
The changes LGTM in general.

@chfsx There is one change in `src/GlobalScreen` (https://github.com/ILIAS-eLearning/ILIAS/pull/5121/files#diff-5769142c345fc3a6bc2301605fe73f7876384f4504e108611a562c99fff70f2b). Do you have any objections?

File: src/GlobalScreen/Scope/Layout/Builder/StandardPageBuilder.php [hunk 1]
```diff
@@ -60,6 +65,16 @@ public function build(PagePartProvider $parts): Page
         $title = $parts->getTitle();
         $short_title = $parts->getShortTitle();
         $view_title = $parts->getViewTitle();
+        $toast_container = $this->ui->factory()->toast()->container();
+
+        foreach (
```
Comment (lscharmer) [inline src/GlobalScreen/Scope/Layout/Builder/StandardPageBuilder.php]: This is just a minor thing but I would suggest to extract something into a variable for clarity.
Something like this:
```php
$toasts = (new ToastsOfNotifications($this->ui->factory(), new ilSetting('notifications'))
    ->transformNotificationsForToasts(
        (new ilNotificationOSDHandler())->getNotificationsForUser($this->user->getId())
    );
foreach ($toasts as $toast) {
    $toast_container = $toast_container->withAdditionalToast($toast);
}
```

**Review by lscharmer — CHANGES_REQUESTED** [review]

**Review by lscharmer — CHANGES_REQUESTED** [review]
Please squash the commits.

**Review by lscharmer — APPROVED** [review]
Thank you :)

File: src/GlobalScreen/Scope/Layout/Builder/StandardPageBuilder.php [hunk 2]
```diff
@@ -60,6 +65,14 @@ public function build(PagePartProvider $parts): Page
         $title = $parts->getTitle();
         $short_title = $parts->getShortTitle();
         $view_title = $parts->getViewTitle();
+        $toast_container = $this->ui->factory()->toast()->container();
+
+        $transformer = new ToastsOfNotifications($this->ui->factory(), new ilSetting('notifications'));
```
Comment (chfsx) [inline src/GlobalScreen/Scope/Layout/Builder/StandardPageBuilder.php]: IMO notifications should be collected the same was as other types of notifications are collected (e.g. see SystemInfos, which has an own Collector and is called here src/GlobalScreen/Scope/Layout/Provider/PagePart/StandardPagePartProvider.php:225). we already have the whole `MainNotificationCollector` thing to collect notifications, would be great we could use this pattern for other types of notifications as well.

**Review by chfsx — CHANGES_REQUESTED** [review]

**Comment by iszmais** [comment]
@chfsx Is there still something missing your requirements?
If yes feel free to leave a comment or otherwise i would appreciate your approval.

Greetings,
@iszmais

**Comment by chfsx** [comment]
Hi @iszmais 

Yes actually my comment from last time is still relevant, but I that you have pushed something to this. We should look at this together. The `StandardPageBuilder` should not get too much dependencies, but especially no new ones. You have addressed this by allowing the `PagePartProvider` to supply the TostContainer. But also in `StandardPagePartProvider` we should keep the dependencies small. For certain things this is not possible (e.g. breadcrumb with ilLocator). In the case of the toasts it would be a good idea to collect the single toasts by ToastProviders and a ToastCollector. Or you can use (maybe better) the existing `NotificationCollector`, like it is done for the `AdministrativeNotifications` (see for example `\ILIAS\GlobalScreen\Scope\Layout\Provider\PagePart\StandardPagePartProvider::getSystemInfos`). Otherwise the toasts will have some kind of special status and it will be difficult to keep track of them in the future.

would you like to have a chat on discord next week?

**Comment by iszmais** [comment]
Hello @chfsx ,

I abstracted the toast provider into an own scope within the global screen. Imo it should not be combined or integrated inside the already exsiting notification scope since the intention of creating the toast was do separete the over the years tangled chat and notfication therefore i think we should start that tangeling again.

Maybe you already can take a look into this i will also fix the tests to it.
I would prefer to present the changes of the existing providers (ChatInvitation, Badge, WhoIsOnline, ...) in seperate PRs wich are depending on this just to make it easier for the target maintainers.

Greetings,
@iszmais

**Comment by iszmais** [comment]
Correlation to https://github.com/iszmais/ILIAS/pull/7

**Comment by chfsx** [comment]
Hi @iszmais 

Thank you for your comments and the discussion in https://github.com/iszmais/ILIAS/pull/7. I pushed you two more commits there that we should still include (one is not directly related to your change, but I noticed it because it didn't have any data on the ToastProviders in the artefact yet). 
Your explanation of the distinction between toasts and notifications convinced me, so we can merge it like that!

**Review by chfsx — APPROVED** [review]
Thanks a lot @iszmais !

**Pull request merged**